README.md
# Dynamic Controls 2

[![Github Releases](https://img.shields.io/github/v/release/anthonyalfimov/SWONIC-Dynamic-Controls?display_name=tag)](https://github.com/anthonyalfimov/SWONIC-Dynamic-Controls/releases)
[![License](https://img.shields.io/github/license/anthonyalfimov/SWONIC-Dynamic-Controls)](https://github.com/anthonyalfimov/SWONIC-Dynamic-Controls/blob/master/LICENSE.txt)
[![Support Email](https://img.shields.io/badge/support-email-blue)](mailto:dynamic.controls@swonic.com)

[![Dynamic Controls Trailer](docs/Banner.png)](https://www.youtube.com/watch?v=NrpUNTRJZtc)

**Dynamic Controls** is a free app for ROLI Dashboard. It turns your ROLI Lightpad Block into a highly customisable MIDI control surface.

Design your own control layout using Buttons, Faders and XY Pads!

*Developed by Anthony Alfimov and Andreas Swoboda.*

---

## App Editions

#### Dynamic Controls
Create up to 16 Buttons, Notes, Faders or XY Pads.

#### Dynamic Controls LE (Light Edition)
Create up to 25 Buttons, Notes or Faders with a limited feature set.

---

## Installation

1. Make sure you have the latest version of [ROLI Dashboard](https://roli.com/products/software/blocks-dashboard) installed. Use ROLI Dashboard to update your Lightpad to the latest firmware.

2. Download the Dynamic Controls installer for your operating system from [swonic.com](https://swonic.com/dynamic-controls/) and run it.

3. Launch ROLI Dashboard and connect your Lightpad to the computer (USB connection is recommended). Select "Dynamic Controls" or "Dynamic Controls LE" to load it onto your Lightpad Block. Press the "Edit" button and select the "Blank" mode to start creating your own MIDI controller!

---

## Support

Dynamic Controls is a non-commercial project.

Available as a free download at [swonic.com](https://swonic.com/dynamic-controls/)

Need any help? Send us a message at dynamic.controls@swonic.com


CHANGELOG.txt
Version 2.1.0 [2022-01-19]
    - Added velocity sensitivity for Note controls, which can be disabled using the new "Fixed Note Velocity" global setting
    - Changed Note control behaviour to retriggeting when activating already held Note controls
    - Added global velocity sensitivity setting for Note controls
    - Added global pressure sensitivity setting for all controls

Version 2.0.5 [2020-10-03]
    - Fixed: saving or loading a mode resets some parameters to their default values in Dashboard 4.3.1

Version 2.0.4 [2020-04-16]
    - Renamed "Overlapping Controls" to "Control Layering"
    - Minor updates to app description

Version 2.0.3 Beta [2020-03-23]
    - Fixed: Mac installer permissions error
    - Updated app details and description

Version 2.0.2 Beta [2020-01-12]
    - Fixed: Setting control colour to black disables this and all subsequent controls

Version 2.0.1 Beta [2019-12-28]
    - Added installers for Mac and Win

Version 2.0.0 Beta [2019-11-17]
    - Overhauled touch processing: improved multitouch support and performance
    - Overhauled appearance: new look for XY pads, global brightness and contrast settings, control shading, highlighting controls for touch and pressure indication
    - Introduced dynamic Dashboard interface layout
    - Added ability to individually assign MIDI channel for each control
    - Added new control type: MIDI Note
    - Added Momentary and Latch behaviours for all control types
    - Added "On Value" parameter for Buttons
    - Improved performance of external MIDI processing
    - Added an option to allow controls (or their parts), covered by other controls, to send MIDI messages
    - Switched to offering two app editions: Dynamic Controls (up to 16 buttons, notes, faders or XY pads with full feature set) and Dynamic Controls LE (up to 25 buttons, notes or faders with limited feature set)

Version 1.1.0 [2019-03-06]
    - Minor fixes

Version 1.0.0 [2019-03-05]
    - Initial release


onChangeScriptGenerator/onChangeScriptGen16.cpp
#include <iostream>
#include <fstream>
#include <string>

// Parameter default values

int getDefaultYAxisCC (int controlIndex) { return 102 + controlIndex; }     // can depend on control index
int getDefaultNote (int controlIndex)    { return 60 + controlIndex; }      // can depend on control index
const int defaultButtonOnValue = 127;
const int defaultNoteVelocity = 127;

// Object lifetime based script decoration

class OnChangeScript
{
public:
    OnChangeScript (std::ofstream& file, const std::string& parameterName)
        : mFile (file)
    {
        mFile
            << "<script onChange=\"" << parameterName << "\">\n"
            << "<![CDATA[\n";
    }

    ~OnChangeScript()
    {
        mFile
            << "]]>\n"
            << "</script>\n";
    }

private:
    std::ofstream& mFile;
};

// Updater for the type-dependent parameters

void updateControlType(std::ofstream& file, int ind, bool areValuesUpdated)
{
//====== BUTTON ================================================================
    file
        << "\tif (ParamControlType" << ind << ".value == 0)\n"
        << "\t{\n";

    // Values
    if (areValuesUpdated)
    {
        file
            << "\t\tParamControlCCB" << ind << ".value = " << defaultButtonOnValue << ";\n";
    }

    // Names and tooltips
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Button CC\";\n"
        << "\t\tParamControlCCA" << ind << ".tooltip = \"Select the MIDI message sent by the control.\\n\\nRange: [0-127]\";\n"

        << "\t\tParamControlCCB" << ind << ".displayName = \"On Value\";\n"
        << "\t\tParamControlCCB" << ind << ".tooltip = \"Set the value sent when the control is activated. When the control is deactivated, [0] is always transmitted.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCB" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";

//====== NOTE ==================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 1)\n"
        << "\t{\n";

    // Values
    if (areValuesUpdated)
    {
        file
            << "\t\tMidiNote" << ind << ".value = " << getDefaultNote (ind) << ";\n"
            << "\t\tParamControlCCB" << ind << ".value = " << defaultNoteVelocity << ";\n";
    }

    // Names and tooltips
    file
        << "\t\tParamControlCCB" << ind << ".displayName = \"Velocity\";\n"
        << "\t\tParamControlCCB" << ind << ".tooltip = \"Set the velocity of the MIDI note sent by the control.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCA" << ind << ".visible = false;\n"
        << "\t\tParamControlCCB" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";

//====== FADER =================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 2)\n"
        << "\t{\n";

    // Names and tooltips
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Fader CC\";\n"
        << "\t\tParamControlCCA" << ind << ".tooltip = \"Select the MIDI message sent by the control.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCB" << ind << ".visible = false;\n";

    file
        << "\t}\n";

//====== XY PAD ================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 3)\n"
        << "\t{\n";

    // Values
    if (areValuesUpdated)
    {
        file
            << "\t\tParamControlCCB" << ind << ".value = " << getDefaultYAxisCC (ind) << ";\n";
    }

    // Names and tooltips
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"X Axis CC\";\n"
        << "\t\tParamControlCCA" << ind << ".tooltip = \"Select the MIDI message sent by the horizontal axis.\\n\\nRange: [0-127]\";\n"

        << "\t\tParamControlCCB" << ind << ".displayName = \"Y Axis CC\";\n"
        << "\t\tParamControlCCB" << ind << ".tooltip = \"Select the MIDI message sent by the vertical axis.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCB" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";
}

int main()
{
    const int controlCount = 16;

    std::string fileName {"OnChangeScripts" + std::to_string (controlCount)
                          + ".littlefoot"};
    std::ofstream outFile;
    outFile.open (fileName);

    // onChange="individualBrightness"
    {
        OnChangeScript script (outFile, "individualBrightness");

        outFile << "\tbrightness.visible = ! individualBrightness.value;\n";
    }

    // onChange="ParamControlType"
    for (int i = 0; i < controlCount; ++i)
    {
        outFile << "\n";
        OnChangeScript script (outFile, "ParamControlType" + std::to_string (i));

        updateControlType(outFile, i, true);
    }

    // onChange="amount"
    {
        outFile << "\n";
        OnChangeScript script (outFile, "amount");

        for (int i = 0; i < controlCount; ++i)
        {
            outFile
                << "\tParamControlType"   << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlColour" << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlMode"   << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamMidiChannel"   << i << ".visible = (" << i << " < amount.value);\n";

            outFile
                << "\tControlPosX"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlPosY"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlWidth"       << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlHeight"      << i << ".visible = (" << i << " < amount.value);\n";

            outFile
                << "\tSendPress"          << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlCCP"         << i << ".visible = (" << i << " < amount.value);\n";

            updateControlType(outFile, i, false);

            outFile << "\n";
        }
    }

    std::cout << "Written to " << fileName << "\n";
}

onChangeScriptGenerator/onChangeScriptGen25.cpp
#include <iostream>
#include <fstream>
#include <string>

// Parameter default values

int getDefaultNote (int controlIndex)    { return 60 + controlIndex; }      // can depend on control index

// Object lifetime based script decoration

class OnChangeScript
{
public:
    OnChangeScript (std::ofstream& file, const std::string& parameterName)
        : mFile (file)
    {
        mFile
            << "<script onChange=\"" << parameterName << "\">\n"
            << "<![CDATA[\n";
    }

    ~OnChangeScript()
    {
        mFile
            << "]]>\n"
            << "</script>\n";
    }

private:
    std::ofstream& mFile;
};

// Updater for the type-dependent parameters

void updateControlType(std::ofstream& file, int ind)
{
//====== BUTTON ================================================================
    file
        << "\tif (ParamControlType" << ind << ".value == 0)\n"
        << "\t{\n";

    // Name
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Button CC\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";

//====== NOTE ==================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 1)\n"
        << "\t{\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCA" << ind << ".visible = false;\n";

    file
        << "\t}\n";

//====== FADER =================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 2)\n"
        << "\t{\n";

    // Name
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Fader CC\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";
}

int main()
{
    const int controlCount = 25;

    std::string fileName {"OnChangeScripts" + std::to_string (controlCount)
                          + ".littlefoot"};
    std::ofstream outFile;
    outFile.open (fileName);

    // onChange="individualBrightness"
    {
        OnChangeScript script (outFile, "individualBrightness");

        outFile << "\tbrightness.visible = ! individualBrightness.value;\n";
    }

    // onChange="ParamControlType"
    for (int i = 0; i < controlCount; ++i)
    {
        outFile << "\n";
        OnChangeScript script (outFile, "ParamControlType" + std::to_string (i));

        updateControlType(outFile, i);
    }

    // onChange="amount"
    {
        outFile << "\n";
        OnChangeScript script (outFile, "amount");

        for (int i = 0; i < controlCount; ++i)
        {
            outFile
                << "\tParamControlType"   << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlColour" << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlMode"   << i << ".visible = (" << i << " < amount.value);\n";

            outFile
                << "\tControlPosX"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlPosY"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlWidth"       << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlHeight"      << i << ".visible = (" << i << " < amount.value);\n";

            updateControlType(outFile, i);

            outFile << "\n";
        }
    }

    std::cout << "Written to " << fileName << "\n";
}

onChangeScriptGenerator/onChangeScriptGenLE.cpp
#include <iostream>
#include <fstream>
#include <string>

// Parameter default values

int getDefaultNote (int controlIndex)    { return 60 + controlIndex; }      // can depend on control index

// Object lifetime based script decoration

class OnChangeScript
{
public:
    OnChangeScript (std::ofstream& file, const std::string& parameterName)
        : mFile (file)
    {
        mFile
            << "<script onChange=\"" << parameterName << "\">\n"
            << "<![CDATA[\n";
    }

    ~OnChangeScript()
    {
        mFile
            << "]]>\n"
            << "</script>\n";
    }

private:
    std::ofstream& mFile;
};

// Updater for the type-dependent parameters

void updateControlType(std::ofstream& file, int ind)
{
//====== BUTTON ================================================================
    file
        << "\tif (ParamControlType" << ind << ".value == 0)\n"
        << "\t{\n";

    // Name
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Button CC\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";

//====== NOTE ==================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 1)\n"
        << "\t{\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCA" << ind << ".visible = false;\n";

    file
        << "\t}\n";

//====== FADER =================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 2)\n"
        << "\t{\n";

    // Name
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Fader CC\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";
}

int main()
{
    const int controlCount = 25;

    std::string fileName {"OnChangeScriptsLE.littlefoot"};
    std::ofstream outFile;
    outFile.open (fileName);

    // onChange="individualBrightness"
    {
        OnChangeScript script (outFile, "individualBrightness");

        outFile << "\tbrightness.visible = ! individualBrightness.value;\n";
    }

    // onChange="fixedVelocity"
    {
        outFile << "\n";
        OnChangeScript script (outFile, "fixedVelocity");

        outFile << "\tvelocitySensitivity.visible = ! fixedVelocity.value;\n";
    }

    // onChange="ParamControlType"
    for (int i = 0; i < controlCount; ++i)
    {
        outFile << "\n";
        OnChangeScript script (outFile, "ParamControlType" + std::to_string (i));

        updateControlType(outFile, i);
    }

    // onChange="amount"
    {
        outFile << "\n";
        OnChangeScript script (outFile, "amount");

        for (int i = 0; i < controlCount; ++i)
        {
            outFile
                << "\tParamControlType"   << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlColour" << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlMode"   << i << ".visible = (" << i << " < amount.value);\n";

            outFile
                << "\tControlPosX"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlPosY"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlWidth"       << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlHeight"      << i << ".visible = (" << i << " < amount.value);\n";

            updateControlType(outFile, i);

            outFile << "\n";
        }
    }

    std::cout << "Written to " << fileName << "\n";
}

onChangeScriptGenerator/onChangeScriptGen.cpp
#include <iostream>
#include <fstream>
#include <string>

// Parameter default values

int getDefaultYAxisCC (int controlIndex) { return 102 + controlIndex; }     // can depend on control index
int getDefaultNote (int controlIndex)    { return 60 + controlIndex; }      // can depend on control index
const int defaultButtonOnValue = 127;
const int defaultNoteVelocity = 127;

// Object lifetime based script decoration

class OnChangeScript
{
public:
    OnChangeScript (std::ofstream& file, const std::string& parameterName)
        : mFile (file)
    {
        mFile
            << "<script onChange=\"" << parameterName << "\">\n"
            << "<![CDATA[\n";
    }

    ~OnChangeScript()
    {
        mFile
            << "]]>\n"
            << "</script>\n";
    }

private:
    std::ofstream& mFile;
};

// Updater for the type-dependent parameters

void updateControlType(std::ofstream& file, int ind, bool areValuesUpdated)
{
//====== BUTTON ================================================================
    file
        << "\tif (ParamControlType" << ind << ".value == 0)\n"
        << "\t{\n";

    // Values
    if (areValuesUpdated)
    {
        file
            << "\t\tParamControlCCB" << ind << ".value = " << defaultButtonOnValue << ";\n";
    }

    // Names and tooltips
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Button CC\";\n"
        << "\t\tParamControlCCA" << ind << ".tooltip = \"Select the MIDI message sent by the control.\\n\\nRange: [0-127]\";\n"

        << "\t\tParamControlCCB" << ind << ".displayName = \"On Value\";\n"
        << "\t\tParamControlCCB" << ind << ".tooltip = \"Set the value sent when the control is activated. When the control is deactivated, [0] is always transmitted.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCB" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";

//====== NOTE ==================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 1)\n"
        << "\t{\n";

    // Values
    if (areValuesUpdated)
    {
        file
            << "\t\tMidiNote" << ind << ".value = " << getDefaultNote (ind) << ";\n"
            << "\t\tParamControlCCB" << ind << ".value = " << defaultNoteVelocity << ";\n";
    }

    // Names and tooltips
    file
        << "\t\tParamControlCCB" << ind << ".displayName = \"Velocity Value\";\n"
        << "\t\tParamControlCCB" << ind << ".tooltip = \"Set the velocity of the MIDI note sent when sliding or gliding to the control, and when striking the control if [Fixed Note Velocity] is enabled.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCA" << ind << ".visible = false;\n"
        << "\t\tParamControlCCB" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";

//====== FADER =================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 2)\n"
        << "\t{\n";

    // Names and tooltips
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"Fader CC\";\n"
        << "\t\tParamControlCCA" << ind << ".tooltip = \"Select the MIDI message sent by the control.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCB" << ind << ".visible = false;\n";

    file
        << "\t}\n";

//====== XY PAD ================================================================
    file
        << "\telse if (ParamControlType" << ind << ".value == 3)\n"
        << "\t{\n";

    // Values
    if (areValuesUpdated)
    {
        file
            << "\t\tParamControlCCB" << ind << ".value = " << getDefaultYAxisCC (ind) << ";\n";
    }

    // Names and tooltips
    file
        << "\t\tParamControlCCA" << ind << ".displayName = \"X Axis CC\";\n"
        << "\t\tParamControlCCA" << ind << ".tooltip = \"Select the MIDI message sent by the horizontal axis.\\n\\nRange: [0-127]\";\n"

        << "\t\tParamControlCCB" << ind << ".displayName = \"Y Axis CC\";\n"
        << "\t\tParamControlCCB" << ind << ".tooltip = \"Select the MIDI message sent by the vertical axis.\\n\\nRange: [0-127]\";\n";

    // Visibility
    file
        << "\t\tMidiNote"        << ind << ".visible = false;\n"
        << "\t\tParamControlCCA" << ind << ".visible = (" << ind << " < amount.value);\n"
        << "\t\tParamControlCCB" << ind << ".visible = (" << ind << " < amount.value);\n";

    file
        << "\t}\n";
}

int main()
{
    const int controlCount = 16;

    std::string fileName {"OnChangeScripts.littlefoot"};
    std::ofstream outFile;
    outFile.open (fileName);

    // onChange="individualBrightness"
    {
        OnChangeScript script (outFile, "individualBrightness");

        outFile << "\tbrightness.visible = ! individualBrightness.value;\n";
    }

    // onChange="fixedVelocity"
    {
        outFile << "\n";
        OnChangeScript script (outFile, "fixedVelocity");

        outFile << "\tvelocitySensitivity.visible = ! fixedVelocity.value;\n";
    }

    // onChange="ParamControlType"
    for (int i = 0; i < controlCount; ++i)
    {
        outFile << "\n";
        OnChangeScript script (outFile, "ParamControlType" + std::to_string (i));

        updateControlType(outFile, i, false);
    }

    // onChange="amount"
    {
        outFile << "\n";
        OnChangeScript script (outFile, "amount");

        for (int i = 0; i < controlCount; ++i)
        {
            outFile
                << "\tParamControlType"   << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlColour" << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamControlMode"   << i << ".visible = (" << i << " < amount.value);\n"
                << "\tParamMidiChannel"   << i << ".visible = (" << i << " < amount.value);\n";

            outFile
                << "\tControlPosX"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlPosY"        << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlWidth"       << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlHeight"      << i << ".visible = (" << i << " < amount.value);\n";

            outFile
                << "\tSendPress"          << i << ".visible = (" << i << " < amount.value);\n"
                << "\tControlCCP"         << i << ".visible = (" << i << " < amount.value);\n";

            updateControlType(outFile, i, false);

            outFile << "\n";
        }
    }

    std::cout << "Written to " << fileName << "\n";
}